File: examples/sim.cpp
#include <Eigen/Core>
#include <chrono>
#include <filesystem>
#include <iostream>
#include <string>

#include "affine_mpc/affine_mpc.hpp"

namespace ampc = affine_mpc;
namespace fs = std::filesystem;


int main()
{
  const auto savedir{fs::temp_directory_path() / "ampc_example"};
  if (fs::exists(savedir / "log.npz")) {
    std::string answer;
    std::cout << "Log file already exists. Overwrite? [Y/n]: ";
    std::getline(std::cin, answer);
    if (answer.empty())
      answer = "y";
    if (std::tolower(answer.front()) != 'y') {
      std::cout << "Exiting." << std::endl;
      return 0;
    }
  }

  const int n{2}, m{1}, T{10}, nc{3};
  const auto param{ampc::Parameterization::linearInterp(T, nc)};
  const ampc::Options opts{
      .use_input_cost = true,
      .slew_control_points = true,
  };
  ampc::CondensedMPC msd_mpc{n, m, param, opts};

  Eigen::Matrix2d A{{0, 1}, {-0.6, -0.1}};
  Eigen::Vector2d B{0, 0.2}, w{0, 0};
  double ts{0.1};
  msd_mpc.setModelContinuous2Discrete(A, B, w, ts);

  Eigen::Vector<double, m> u_min{0}, u_max{3}, slew{1};
  msd_mpc.setInputLimits(u_min, u_max);
  msd_mpc.setSlewRate(slew);

  Eigen::Vector<double, n> Q_diag{1, 0.1};
  Eigen::Vector<double, m> R_diag{1e-5};
  msd_mpc.setWeights(Q_diag, R_diag);

  Eigen::Vector2d x_ref{1, 0};
  Eigen::Vector<double, m> u_ref{3.0}; // 3N force holds mass at 1m
  msd_mpc.setReferenceState(x_ref);
  msd_mpc.setReferenceInput(u_ref);

  if (!msd_mpc.initializeSolver()) {
    std::cerr << "Failed to initialize solver\n";
    return 1;
  }

  const int pred_stride{1}; // 1 logs every step of each solve horizon
  ampc::MPCLogger logger{&msd_mpc, savedir, ts, pred_stride};
  logger.addMetadata("example_name", "mass_spring_damper");

  Eigen::Vector2d xk{0, 0};
  Eigen::Matrix<double, m, 1> uk;

  double tf{10};
  affine_mpc::SolveStatus status;
  for (double t{0}; t < tf; t += ts) {
    auto start = std::chrono::high_resolution_clock::now();
    status = msd_mpc.solve(xk);
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    if (status != ampc::SolveStatus::Success)
      std::cout << "Solver status: " << status << std::endl;

    msd_mpc.getNextInput(uk);
    logger.logStep(t, xk, elapsed.count());

    msd_mpc.propagateModel(xk, uk, xk);
  }

  std::cout << "Log file written to " << savedir / "log.npz" << std::endl;

  return 0;
}

File: examples/sim.py
# std lib
from pathlib import Path
import tempfile
from time import perf_counter

# 3rd party
import numpy as np

# local
import affine_mpc as ampc


def plot_example():
    if __name__ == "__main__":
        import plot_sim
    else:
        from . import plot_sim

    plot_sim.main()


def run_example():
    tmp = Path(tempfile.gettempdir())
    save_dir = tmp / "ampc_example"
    if (save_dir / "log.npz").exists():
        answer = input("Log file already exists. Overwrite? [Y/n]: ")
        if len(answer) == 0:
            answer = "y"
        if not answer.lower().startswith("y"):
            print("Not overwriting.")
            return

    msd_mpc = ampc.CondensedMPC(
        state_dim=2,
        input_dim=1,
        param=ampc.Parameterization.linearInterp(
            horizon_steps=10, num_control_points=3
        ),
        opts=ampc.Options(use_input_cost=True, slew_control_points=True),
    )

    A = np.array([[0, 1], [-0.6, -0.1]])
    B = np.array([0, 0.2])
    w = np.zeros(2)
    ts = 0.1
    msd_mpc.setModelContinuous2Discrete(A, B, w, ts)

    u_min = np.zeros(1)
    u_max = np.array([3.0])
    msd_mpc.setInputLimits(u_min, u_max)

    slew = np.ones(1)
    msd_mpc.setSlewRate(slew)

    Q_diag = np.array([1, 0.1])
    R_diag = np.array([1e-5])
    msd_mpc.setWeights(Q_diag, R_diag)

    x_ref = np.array([1.0, 0])
    u_ref = np.array([3.0])  # 3N force holds mass at 1m
    msd_mpc.setReferenceState(x_ref)
    msd_mpc.setReferenceInput(u_ref)

    if not msd_mpc.initializeSolver():
        raise RuntimeError("Failed to initialize solver")

    pred_stride = 1  # 1 logs every step of each solve horizon
    logger = ampc.MPCLogger(msd_mpc, save_dir, ts, pred_stride)
    logger.addMetadata("example_name", "mass_spring_damper")

    xk = np.zeros(2)

    t = 0.0
    tf = 10.0
    while t < tf:
        start = perf_counter()
        status = msd_mpc.solve(xk)
        elapsed = perf_counter() - start

        if status != ampc.SolveStatus.Success:
            print("Solver status:", status)

        uk = msd_mpc.getNextInput()
        logger.logStep(t, xk, elapsed)

        # Simulate system
        xk = msd_mpc.propagateModel(xk, uk)
        t += ts

    print("Log file written to", save_dir / "log.npz")


def main():
    run_example()
    plot_example()


if __name__ == "__main__":
    main()
